src/sage/data_structures/pairing_heap.pyx

```diff
@@ -0,0 +1,1501 @@
+        sage: P.top()
+        ('abc', 2)
+        sage: P.decrease(1, 1)
+        sage: P.top()
+        (1, 1)
+        sage: P.pop()
+        sage: P.top()
+        ('abc', 2)
+
+        sage: P = PairingHeap_of_n_hashables(5)
+        sage: P.push(1, (2, 3))
+        sage: P.push('a', (2, 2))
+        sage: P.push('b', (3, 3))
+        sage: P.push('c', (2, 1))
+        sage: P.top()
+        ('c', (2, 1))
+        sage: P.push(Graph(2, immutable=True), (1, 7))
+        sage: P.top()
+        (Graph on 2 vertices, (1, 7))
+        sage: P.decrease('b', (1, 5))
+        sage: P.top()
+        ('b', (1, 5))
+
+    TESTS::
+
+        sage: from sage.data_structures.pairing_heap import PairingHeap_of_n_hashables
+        sage: P = PairingHeap_of_n_hashables(0)
+        Traceback (most recent call last):
+        ...
+        ValueError: the capacity of the heap must be strictly positive
+        sage: P = PairingHeap_of_n_hashables(1); P
+        PairingHeap_of_n_hashables: capacity 1, size 0
+        sage: P.push(11, 3)
+        sage: P.push(12, 4)
+        Traceback (most recent call last):
+        ...
+        ValueError: the heap is full
+    """
+    def __init__(self, size_t n):
+        r"""
+        Construct the ``PairingHeap_of_n_hashables``.
+
+        This pairing heap has a maximum capacity of `n` items and each item is a
+        hashable object.
+
+        INPUT:
+
+        - ``n`` -- strictly positive integer; the maximum number of items in the heap
+
+        EXAMPLES::
+
+            sage: from sage.data_structures.pairing_heap import PairingHeap_of_n_hashables
+            sage: P = PairingHeap_of_n_hashables(2); P
+            PairingHeap_of_n_hashables: capacity 2, size 0
+            sage: P.push(1, 2)
+            sage: P
+            PairingHeap_of_n_hashables: capacity 2, size 1
+            sage: P.push(2, 3)
+            sage: P
+            PairingHeap_of_n_hashables: capacity 2, size 2
+            sage: P.full()
+            True
+            sage: P.push(10, 1)
+            Traceback (most recent call last):
+            ...
+            ValueError: the heap is full
+            sage: P.pop()
+            sage: P
+            PairingHeap_of_n_hashables: capacity 2, size 1
+            sage: P.push(10, 1)
+
+    TESTS::
+
+            sage: P = PairingHeap_of_n_hashables(0)
+            Traceback (most recent call last):
+            ...
+            ValueError: the capacity of the heap must be strictly positive
+            sage: P = PairingHeap_of_n_hashables(1); P
+            PairingHeap_of_n_hashables: capacity 1, size 0
+        """
+        if not n:
+            raise ValueError("the capacity of the heap must be strictly positive")
+        self.name = "PairingHeap_of_n_hashables"
+        self.n = n
+        self.root = NULL
+        self.nodes = <PairingHeapNode *>check_allocarray(n, sizeof(PairingHeapNode))
+        bitset_init(self.active, n)
+        bitset_clear(self.active)
```

Comment:
same remark as in the other class: needed ?

Reply:
not needed.